app/models/reservation.rb

```diff
@@ -17,4 +17,24 @@ class Reservation < ApplicationRecord # :nodoc:
   validates :checkin_at, presence: true
   validates :checkout_at, presence: true
   validates :book_code, presence: true, uniqueness: true
+  validate :room_must_be_available, if: -> { room_id.present? || checkin_at.present? || checkout_at.present? }
+
+  # Scopes
+  scope :avoid_overlapping_reservations, lambda { |checkin_at, checkout_at, room_id, reservation_id = nil|
+    where('checkin_at < ? AND checkout_at > ?', checkout_at, checkin_at)
+      .where(room_id:)
+      .where.not(id: reservation_id)
+  }
```

Comment:
I think the logic is incomplete if you are looking for overlapping reservations, and also you can re-order it to be more efficient and readable:

1) move`.where(room_id:).where.not(id: reservation_id)` first will quickly filter down to specific room and exclude same reservation. This can already speed up the query.

2) I think the overlapping logic is not complete. The overlapping cases could be:
- new reservation checkin time between existing reservation checkin and checkout time; OR
- new reservation checkout time between existing reservation checkin and checkout time; OR
- new reservation checkin AND checkout time between (within) existing reservation checkin and checkout time; OR
- new reservation checkin and checkout time encompass the existing reservation checkin and checkout time (ie. existing reservation checkin and checkout time is within the new reservation timeframe)